src/main/java/org/prebid/server/cookie/UidsCookieService.java

```diff
@@ -250,43 +275,53 @@ private static UidsCookie removeExpiredUids(UidsCookie uidsCookie) {
         return updatedCookie;
     }
 
-    private UidsCookieUpdateResult updateUidsCookieByPriority(UidsCookie uidsCookie, String familyName, String uid) {
-        final UidsCookie updatedCookie = uidsCookie.updateUid(familyName, uid);
-        if (!cookieExceededMaxLength(updatedCookie)) {
-            return UidsCookieUpdateResult.updated(updatedCookie);
-        }
+    public List<Cookie> splitUidsIntoCookies(UidsCookie uidsCookie) {
+        final Uids cookieUids = uidsCookie.getCookieUids();
+        final Map<String, UidWithExpiry> uids = cookieUids.getUids();
+        final boolean hasOptout = !uidsCookie.allowsSync();
 
-        if (!prioritizedCoopSyncProvider.hasPrioritizedBidders()
-                || prioritizedCoopSyncProvider.isPrioritizedFamily(familyName)) {
-            return UidsCookieUpdateResult.updated(trimToLimit(updatedCookie));
-        } else {
-            metrics.updateUserSyncSizeBlockedMetric(familyName);
-            return UidsCookieUpdateResult.unaltered(uidsCookie);
-        }
-    }
+        final Iterator<String> cookieFamilies = cookieFamilyNamesByDescPriorityAndExpiration(uidsCookie);
+        final List<Cookie> splitCookies = new ArrayList<>();
 
-    private boolean cookieExceededMaxLength(UidsCookie uidsCookie) {
-        return maxCookieSizeBytes > 0 && cookieBytesLength(uidsCookie) > maxCookieSizeBytes;
-    }
+        int uidsIndex = 0;
+        String nextCookieFamily = null;
 
-    private UidsCookie trimToLimit(UidsCookie uidsCookie) {
-        if (!cookieExceededMaxLength(uidsCookie)) {
-            return uidsCookie;
-        }
+        while (uidsIndex < numberOfUidCookies) {
+            final String uidsName = uidsIndex == 0 ? COOKIE_NAME : COOKIE_NAME_FORMAT.formatted(uidsIndex + 1);
+            UidsCookie tempUidsCookie = new UidsCookie(
+                    Uids.builder().uids(new HashMap<>()).optout(hasOptout).build(),
+                    mapper);
 
-        UidsCookie trimmedUids = uidsCookie;
-        final Iterator<String> familyToRemoveIterator = cookieFamilyNamesByAscendingPriority(uidsCookie);
+            while (nextCookieFamily != null || cookieFamilies.hasNext()) {
+                nextCookieFamily = nextCookieFamily == null ? cookieFamilies.next() : nextCookieFamily;
+                tempUidsCookie = tempUidsCookie.updateUid(nextCookieFamily, uids.get(nextCookieFamily));
+                if (cookieExceededMaxLength(uidsName, tempUidsCookie)) {
```

Comment:
I don't like that we have to encode cookies to ensure they don't exceed the size. We need to try to optimize the validation process.

We need to discuss it. `Instrumentation.getObjectSize` as a first option